### PR TITLE
Fix clippy warnings.

### PR DESCRIPTION
**Pull request by Frostie314159**
This also removes multiple uses of transmute, which have safe equivalents. All tests pass and clippy is happy.

### CODE REVIEW

File: src/ctx.rs [hunk 1]
```diff
@@ -621,13 +619,7 @@ macro_rules! into_ctx_float_impl {
             #[inline]
             fn into_ctx(self, dst: &mut [u8], le: Endian) {
                 assert!(dst.len() >= $size);
-                write_into!(
-                    signed_to_unsigned!($typ),
-                    $size,
-                    transmute::<$typ, signed_to_unsigned!($typ)>(self),
```
Comment (m4b) [inline src/ctx.rs:627]: just want to be sure this doesn't affect semantics here, the `signed_to_unsigned!($typ)` in the transmute param is ok to replace with just to_bits(), yes?  (I haven't reviewed this code in literally years and years, so my memory for why this stuff is even here is foggy)

**Review by m4b — APPROVED** [review]
this looks like a great initial cleanup, thank you so much! only question was on the one transmute, but it looks fine to me. otherwise i think we have to merge your other PR and fix ci in it before we merge this one?

File: src/ctx.rs (hunk 1 above)
Comment (Frostie314159) [reply src/ctx.rs:627]: The [docs](https://doc.rust-lang.org/std/primitive.f32.html#method.to_bits) for this method show, that the unsigned version is always returned.

File: src/ctx.rs [hunk 2]
```diff
@@ -849,6 +841,28 @@ impl TryIntoCtx for CString {
         self.as_c_str().try_into_ctx(dst, ())
     }
 }
+impl<'a, const N: usize> TryFromCtx<'a> for [u8; N] {
```
[the remark this answers is not in the document]
Comment (Frostie314159) [reply src/ctx.rs:844]: I'm gonna revert those commits, since they weren't intended to be here in the first place.

**Comment by nyurik** [comment]
I fixed the CI - MSRV in #90

**Comment by nyurik** [comment]
I fixed the CI - MSRV in #90

File: src/ctx.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Frostie314159) [reply src/ctx.rs:843]: Fixed in latest commit.

**Comment by m4b** [comment]
so if you're having trouble fixing this, you can do:
```
git checkout master
git pull
git checkout clippy-fix # assumes this is your branch name
git rebase master
# you will likely have to manually fix the conflict markers (choose the 0x4a case)
git add src/lib.rs
git rebase --continue
git push --force # force push the rebased change to your branch 
```

**Comment by nyurik** [comment]
@m4b and @Frostie314159 I created a PR for this PR - fixing merge conflicts. It should be merged in before merging this PR - https://github.com/Frostie314159/scroll/pull/1

**Comment by m4b** [comment]
@nyurik it still says there is a merge conflict?

**Comment by m4b** [comment]
Also, i'm not sure what's going on, but i'm having trouble reviewing the "final changes" of this PR; when i choose to review all commits, only 3 files are changed, and some changes that were supposedly removed (the try_from_ctx impl for `[u8; N]`) is still present, and all the transmute and safety fixes are missing. :confused: 
review_screenshot

I'd expect reviewing all changes to show me what i'm going to merge, maybe i haven't done this in a while and i my memory doesn't serve, but i don't remember github presenting stacks of commits in this manner? (i remember being able to view what he entire set of changes introduced, which is what i want to see, and which is what would be merged into master)

**Comment by nyurik** [comment]
@m4b those fixes were also done and merged in by my earlier PR.  This PR only has the stuff that my (also clippy-inspired PR) did not have. The diff tool (add `/compare` to your repo) will help us explore this:

Main branch has two commits: f5beda1c9 and 30f49c7bb. The second commit is mostly irrelevant (format inlining), so comparing with it is pointless because it will just increase the noise. The first commit on the other hand is the one that contains a lot of the changes that were also done by this commit.

You can see it by comparing it with 62c5fcf90325f: 
 https://github.com/m4b/scroll/compare/f5beda1c9..62c5fcf -- this is the commit right before my PR was merged into the @Frostie314159 fork.

Now the very next commit - caf60bb: 
* this is what changed between @Frostie314159 commits: https://github.com/m4b/scroll/compare/62c5fcf..caf60bb
* this is the new diff with the primary branch https://github.com/m4b/scroll/compare/f5beda1c9..caf60bb

Next is the merge commit --  eac19ef -- merge commit is a noop for us, so same results as before
* this is what changed between @Frostie314159 commits: https://github.com/m4b/scroll/compare/caf60bb..eac19ef
* this is the new diff with the primary branch https://github.com/m4b/scroll/compare/f5beda1c9..eac19ef

So in short - all transmute fixes are already in the main branch.

**Comment by nyurik** [comment]
P.S. the conflict is tiny - I fixed it in https://github.com/Frostie314159/scroll/pull/2 -- this is again just a simple PR that was created with `git merge upstream/master` and fixing conflicts.

-- this PR modifies two lines: [this](https://github.com/m4b/scroll/pull/88/files#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R388) and [this](https://github.com/m4b/scroll/pull/88/files#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R397) - both in `src/lib.rs` - and my last PR modifies `println!("1 {:?}", &hello_world);` which is right above those lines. So git requires manual check when two adjacent lines are modified by two separate commits.

**Comment by Frostie314159** [comment]
Thank you so much!

File: src/lib.rs [hunk 3]
```diff
@@ -486,9 +483,9 @@ mod tests {
         let bytes_from: [u8; 8] = [1, 2, 3, 4, 5, 6, 7, 8];
         let bytes_to = &mut bytes_to[..];
         let bytes_from = &bytes_from[..];
-        for i in 0..bytes_from.len() {
-            bytes_to[i] = bytes_from.pread(i).unwrap();
-        }
+        bytes_to.iter_mut().enumerate().for_each(|(i, item)| {
+            *item = bytes_from.pread(i).unwrap();
+        });
```
Comment (nyurik) [inline src/lib.rs:488]: i think this is a bit hard to read - iterating over the "target" as mutable.  I would suggest a simpler variant (i didn't test it, might not even compile) -- iterate over source values, pushing them into the output array.


```suggestion
        for (idx, value) in bytes_from.iter().enumerate() {
            bytes_to[i] = value.unwrap();
        });
```

File: src/lib.rs [hunk 4]
```diff
@@ -580,10 +577,10 @@ mod tests {
         let bytes_from: [u8; 8] = [1, 2, 3, 4, 5, 6, 7, 8];
         let bytes_to = &mut bytes_to[..];
         let bytes_from = &bytes_from[..];
-        let mut offset = &mut 0;
-        for i in 0..bytes_from.len() {
-            bytes_to[i] = bytes_from.gread(&mut offset).unwrap();
-        }
+        let offset = &mut 0;
+        bytes_to.iter_mut().for_each(|item| {
```
Comment (nyurik) [inline src/lib.rs:581]: same as above

File: tests/api.rs [hunk 5]
```diff
@@ -136,6 +137,11 @@ impl<'a> Segments<'a> {
         Ok(sections)
     }
 }
+impl<'a> Default for Segments<'a> {
```
Comment (nyurik) [inline tests/api.rs:140]: Why is this needed?  You can simply add `#[derive(Default)]` on the `Segments` struct, and make `new()` call `Self::default()`.  Also, I am not too certain if both `default()` and `new()` are needed. I think cleaner interface would be to have just one. This way if later you need `new(some_init_value)`, it can be easily added.

File: src/lib.rs (hunk 4 above)
Comment (nyurik) [reply src/lib.rs:581]: although this might need to be different - not certain about this one

Comment (Frostie314159) [reply src/lib.rs:581]: This was the clippy suggestion.

File: tests/api.rs (hunk 5 above)
Comment (Frostie314159) [reply tests/api.rs:140]: Good idea, I can get to this tomorrow.

**Comment by m4b** [comment]
i believe all these fixes have been merged, if not we should do this in another PR since this has conflicts, etc.
